crates/flux-fhir-analysis/src/wf/sortck.rs

```diff
@@ -270,8 +279,42 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
                     _ => Err(self.emit_field_not_found(&sort, *fld)),
                 }
             }
-            // (VR): Pretty sure we don't synthesize these?
-            _ => span_bug!(expr.span, "unexpected expr `{expr:?}` in synth expr"),
+            fhir::ExprKind::Constructor(Some(path), field_exprs, spread) => {
+                // if we have S then we can synthesize otherwise we fail
+                // first get the sort based on the path - for example S { ... } => S
+                // and we should expect sort to be a struct or enum app
+                let path_def_id = match path.res {
+                    ExprRes::Struct(def_id) | ExprRes::Enum(def_id) => def_id,
+                    _ => span_bug!(expr.span, "unexpected path in constructor"),
+                };
+                let sort_def = self
+                    .genv
+                    .adt_sort_def_of(path_def_id)
+                    .map_err(|e| self.emit_err(e))?;
+                // generate fresh inferred sorts for each param
+                let fresh_args = (0..sort_def.param_count())
+                    .map(|_| self.next_sort_var())
+                    .collect_vec();
+                let sort_by_name = sort_def.sort_by_field_name(&fresh_args);
+                let sort = rty::Sort::App(rty::SortCtor::Adt(sort_def.clone()), fresh_args.into());
+
+                // check the shape
+                self.check_constructor_shape(expr.span, &sort_by_name, field_exprs, spread, &sort)?;
+                // synth sort
+                for expr in *field_exprs {
+                    // check each expression against the sort
+                    // which will unify inferred variables
+                    if let Some(sort) = sort_by_name.get(&expr.ident.name) {
+                        self.check_expr(&expr.expr, sort)?;
+                    }
+                }
+                // check the spread against the sort
+                if let Some(spread) = spread {
+                    self.check_expr(&spread.expr, &sort)?;
+                }
```

Comment:
This is the same as `check_field_expr`, no? Maybe we don't need `check_constructor_shape` and could call the old `check_field_expr`.

Reply:
Yea it is - good point.